ethcore/sync/src/chain/mod.rs

```diff
@@ -855,14 +853,29 @@ impl ChainSync {
 	fn collect_blocks(&mut self, io: &mut SyncIo, block_set: BlockSet) {
 		match block_set {
 			BlockSet::NewBlocks => {
-				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == Err(DownloaderImportError::Invalid) {
+				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == DownloadAction::Reset {
+					trace!(target: "sync", "Restarting NewBlocks download");
 					self.restart(io);
 				}
 			},
 			BlockSet::OldBlocks => {
-				if self.old_blocks.as_mut().map_or(false, |downloader| { downloader.collect_blocks(io, false) == Err(DownloaderImportError::Invalid) }) {
-					self.restart(io);
```

Comment:
Should it restart in the case of a bad old block?

Reply:
Yes and that behaviour is retained.

Reply:
Now it resets peer requests and the downloader state instead of calling `restart()`, so it looks like the behavior after an invalid ancient block has changed. But I think the new logic is better (it doesn't mess with the new blocks downloader state).

Reply:
You're right, I misunderstood that comment last night. Yes that was intentional to stop the NewBlocks downloader being reset. I think that similarly the OldBlocks download perhaps shouldn't be reset (via restart) when we get a bad block for NewBlocks. Plenty of improvements still to make here!